proof/invariant-abstract/DetSchedInvs_AI.thy

```diff
@@ -581,26 +581,30 @@ end
 
 \<comment> \<open>Project threads from the kernel heap\<close>
 
-definition tcb_of :: "kernel_object \<rightharpoonup> tcb" where
-  "tcb_of ko \<equiv> case ko of TCB tcb \<Rightarrow> Some tcb | _ \<Rightarrow> None"
-
-lemma tcb_of_simps:
+primrec tcb_of :: "Structures_A.kernel_object \<rightharpoonup> tcb" where
   "tcb_of (TCB tcb) = Some tcb"
-  "tcb_of (SchedContext sc n) = None"
-  "tcb_of (CNode n cnode) = None"
-  "tcb_of (Endpoint ep) = None"
-  "tcb_of (Notification ntfn) = None"
-  "tcb_of (Reply reply) = None"
-  "tcb_of (ArchObj aobj) = None"
-  by (auto simp: tcb_of_def)
+| "tcb_of (Endpoint _) = None"
+| "tcb_of (Notification _) = None"
+| "tcb_of (CNode _ _) = None"
+| "tcb_of (Reply _) = None"
+| "tcb_of (SchedContext _ _) = None"
+| "tcb_of (ArchObj _) = None"
+
+lemma tcb_of_SomeD[dest!]:
+  "tcb_of ko = Some obj \<Longrightarrow> ko = TCB obj"
+  by (cases ko; simp)
+
+lemma tcb_of_NoneD[simp]:
```

Comment:
I don't think this is a good simp rule, the conclusion is not in simp normal form and the left-hand side might often be not true, so if the preprocessed conclusion fits often, this will be applied often and fail often, i.e. slow down proofs where it doesn't work.

But: if we can make it an equality instead of `==>` similar to my comment on `tcb_of_SomeD`, then `tcb_of ko = None` becomes the lhs of the rewrite rule, which would be perfectly nice for a `[simp]` rule.

Reply:
This might not quite work as an automatic simp rule, I am afraid. For the `Some` case, the equality in the conclusion allows us to retrieve the equality in the LHS, so we don’t lose anything. But for this `None` case, we won’t be able to use the LHS equality once this simp rule triggers. This leaves us with stray `tcb_of ko` in other expressions which we’d want to rewrite to `None`.

Reply:
Hm, yes, you're right. There's no good mechanism for saying this should only be applied when nothing else works. Still not sure it should be globally `simp`. Do you have an example goal where this rule applies? (Just trying to figure out how the pattern works in practise)

Reply:
I now pushed a version with all the `xxx_of_None` in the simp set. There are two `auto` proofs in DetSched files that get broken by having `tcb_of_None` in the simp set.The subgoal for one of them is like this (it's quite messy):
```
 ⟦P (the (map_option (r_time ∘ hd ∘ scrc_refills)
                 (case case tcb_of x2 of None ⇒ None
                       | Some x ⇒ (id ∘ tcb_sched_context) x of
                  None ⇒ None
                  | Some x ⇒
                      map_project sc_refill_cfg_of
                       (λsa. case kheap s sa of None ⇒ None
                             | Some x ⇒ sc_of x)
                       x)));
        ∀x. sc_refills (f x) = sc_refills x;
        kheap s sc_ptr = Some (SchedContext x xa);
        a_type (SchedContext x xa) = a_type (SchedContext (f x) xa);
        kheap s t = Some x2; ∀tcb. x2 ≠ TCB tcb⟧
       ⟹ P (the None)
       ```
It's only `tcb_of_None` causing issues, but it is probably because others aren't used as much. I didn't fix Refine, so we might see more breakage there.

Reply:
I have a version that I haven’t pushed that has the equalities for `Some` cases in the simp set while those for `None` not in the simp set, and this seems to work well (doesn’t break any proof). I would be happy if we can settle for this version.

Taking the `Some` equalities out of the simp set does break the proof (`sc_of_Some` is needed in many places in Refine).

Reply:
I pushed the above version as a separate commit. If there is no further comments, I would like an approval so that I can merge this soon. I will squash and update the commit messages before merging.

Reply:
I'm happy with this part. Still need to look through the rest. Should be done soon, though.